examples/dqn/dqn_cartpole.py

```diff
@@ -0,0 +1,183 @@
+from torchrl.envs import ExplorationType, set_exploration_type
+from torchrl.modules import EGreedyWrapper
+from torchrl.objectives import DQNLoss, HardUpdate
+from torchrl.record.loggers import generate_exp_name, get_logger
+from utils_cartpole import eval_model, make_dqn_model, make_env
+
+
+@hydra.main(config_path=".", config_name="config_cartpole", version_base="1.1")
+def main(cfg: "DictConfig"):  # noqa: F821
+
+    device = "cpu" if not torch.cuda.device_count() else "cuda"
+
+    # Make the components
+    model = make_dqn_model(cfg.env.env_name)
+    model_explore = EGreedyWrapper(
+        policy=model,
+        annealing_num_steps=cfg.collector.annealing_frames,
+        eps_init=cfg.collector.eps_start,
+        eps_end=cfg.collector.eps_end,
+    ).to(device)
+
+    # Create the collector
+    collector = SyncDataCollector(
+        create_env_fn=make_env(cfg.env.env_name, device),
+        policy=model_explore,
+        frames_per_batch=cfg.collector.frames_per_batch,
+        total_frames=cfg.collector.total_frames,
+        device=device,
+        storing_device=device,
+        max_frames_per_traj=-1,
+        init_random_frames=cfg.collector.init_random_frames,
+    )
+
+    # Create the replay buffer
+    replay_buffer = TensorDictReplayBuffer(
+        pin_memory=False,
+        prefetch=3,
+        storage=LazyTensorStorage(
+            max_size=cfg.buffer.buffer_size,
+            device=device,
+        ),
+        batch_size=cfg.buffer.batch_size,
+    )
+
+    # Create the loss module
+    loss_module = DQNLoss(
+        value_network=model,
+        gamma=cfg.loss.gamma,
+        loss_function="l2",
+        delay_value=True,
+    )
+    loss_module.make_value_estimator(gamma=cfg.loss.gamma)
+    target_net_updater = HardUpdate(
+        loss_module, value_network_update_interval=cfg.loss.hard_update_freq
+    )
+
+    # Create the optimizer
+    optimizer = torch.optim.Adam(loss_module.parameters(), lr=cfg.optim.lr)
+
+    # Create the logger
+    logger = None
+    if cfg.logger.backend:
+        exp_name = generate_exp_name("DQN", f"CartPole_{cfg.env.env_name}")
+        logger = get_logger(cfg.logger.backend, logger_name="dqn", experiment_name=exp_name)
+
+    # Create the test environment
+    test_env = make_env(cfg.env.env_name, device)
+
+    # Main loop
+    collected_frames = 0
+    start_time = time.time()
+    num_updates = cfg.loss.num_updates
+    batch_size = cfg.buffer.batch_size
+    test_interval = cfg.logger.test_interval
+    num_test_episodes = cfg.logger.num_test_episodes
+    frames_per_batch = cfg.collector.frames_per_batch
+    pbar = tqdm.tqdm(total=cfg.collector.total_frames)
+    sampling_start = time.time()
+
+    for data in collector:
+
+        log_info = {}
+        sampling_time = time.time() - sampling_start
+        pbar.update(data.numel())
+        data = data.reshape(-1)
+        current_frames = data.numel()
+        replay_buffer.extend(data.to(device))
```

Comment:
seems weird that we are sending the data to device before putting it in the replay buffer.
Usually the buffer is on disk/ram and the data on cuda, not the other way around